crates/bevy_ecs/src/observer/runner.rs

```diff
@@ -64,37 +68,23 @@ pub(super) fn observer_system_runner<E: Event, B: Bundle, S: ObserverSystem<E, B
         #[cfg(feature = "hotpatching")]
         (*system).refresh_hotpatch();
 
-        match (*system).validate_param_unsafe(world) {
-            Ok(()) => {
-                if let Err(err) = (*system).run_unsafe(trigger, world) {
-                    let handler = state
-                        .error_handler
-                        .unwrap_or_else(|| world.default_error_handler());
-                    handler(
-                        err,
-                        ErrorContext::Observer {
-                            name: (*system).name(),
-                            last_run: (*system).get_last_run(),
-                        },
-                    );
-                };
-                (*system).queue_deferred(world.into_deferred());
-            }
-            Err(e) => {
-                if !e.skipped {
-                    let handler = state
-                        .error_handler
-                        .unwrap_or_else(|| world.default_error_handler());
-                    handler(
-                        e.into(),
-                        ErrorContext::Observer {
-                            name: (*system).name(),
-                            last_run: (*system).get_last_run(),
-                        },
-                    );
-                }
-            }
-        }
+        if let Err(RunSystemError::Failed(err)) = (*system)
```

Comment:
Idly interested in benchmarks of this. Removing the branch here should be net positive.